# SVM with Kernel: RBF

In [8]:
# import the necessary packages
import time
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import classification_report,confusion_matrix
from sklearn.metrics import precision_score,f1_score

In [3]:
df = pd.read_csv('mnist.csv')       # read the csv file

df_feat = df.drop('label',axis=1)   # drop the label column

label_y = df['label']               # label column

In [5]:
x_train, x_test, y_train, y_test = train_test_split(df_feat,label_y,test_size=0.5,random_state=101)

start = time.time()
model = SVC(kernel='rbf',gamma=1,C=0.1)                         # rbf kernel gamma=1 and C=0.1
model.fit(x_train,y_train)
pred = model.predict(x_test)
print("Gamma: 1 , C: 0.1")
report = classification_report(y_test,pred,output_dict=True)    # classification report
print("Confusion Matrix: ",confusion_matrix(y_test,pred))       # confusion matrix
print("Accuracy: ",report['accuracy'])
end = time.time()
print("--- %s mins %s secs ---" % ((end - start) // 60, (end - start) % 60))


Gamma: 1 , C: 0.1
Confusion Matrix:  [[   0 2038    0    0    0    0    0    0    0    0]
 [   0 2345    0    0    0    0    0    0    0    0]
 [   0 2092    0    0    0    0    0    0    0    0]
 [   0 2154    0    0    0    0    0    0    0    0]
 [   0 2052    0    0    0    0    0    0    0    0]
 [   0 1886    0    0    0    0    0    0    0    0]
 [   0 2061    0    0    0    0    0    0    0    0]
 [   0 2188    0    0    0    0    0    0    0    0]
 [   0 2059    0    0    0    0    0    0    0    0]
 [   0 2125    0    0    0    0    0    0    0    0]]
Accuracy:  0.11166666666666666
--- 19.0 mins 43.02353620529175 secs ---


c:\Users\Dhruv\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
c:\Users\Dhruv\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
c:\Users\Dhruv\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [6]:
start = time.time()
model = SVC(kernel='rbf',gamma=0.1,C=1)                         # gamma = 0.1 , C = 1
model.fit(x_train,y_train)
pred = model.predict(x_test)
print("Gamma: 0.1 , C: 1")
report = classification_report(y_test,pred,output_dict=True)    # classification report
print("Confusion Matrix: ",confusion_matrix(y_test,pred))       # confusion matrix
print("Accuracy: ",report['accuracy'])
end = time.time()
print("--- %s mins %s secs ---" % ((end - start) // 60, (end - start) % 60))

Gamma: 0.1 , C: 1
Confusion Matrix:  [[   0 2038    0    0    0    0    0    0    0    0]
 [   0 2345    0    0    0    0    0    0    0    0]
 [   0 2092    0    0    0    0    0    0    0    0]
 [   0 2154    0    0    0    0    0    0    0    0]
 [   0 2052    0    0    0    0    0    0    0    0]
 [   0 1886    0    0    0    0    0    0    0    0]
 [   0 2061    0    0    0    0    0    0    0    0]
 [   0 2188    0    0    0    0    0    0    0    0]
 [   0 2059    0    0    0    0    0    0    0    0]
 [   0 2125    0    0    0    0    0    0    0    0]]
Accuracy:  0.11166666666666666
--- 17.0 mins 30.035959482192993 secs ---


c:\Users\Dhruv\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
c:\Users\Dhruv\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
c:\Users\Dhruv\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


This kernel performs the worst for this dataset with the highest inaccuracy (Accuracy = 0.11) and longest time taken to train around 19 and 17 mins.